cpp/nccl_comms/src/cuML_comms_nccl_impl.cpp

```diff
@@ -0,0 +1,445 @@
+ * cuml.
+ *
+ * For instance, nccl-py can be used to bootstrap a ncclComm_t before it is
+ * used to construct a cuml comms instance. UCX endpoints can be bootstrapped
+ * in Python as well, before being used to construct a cuML comms instance.
+ */
+void inject_comms(cumlHandle &handle, ncclComm_t comm, ucp_worker_h ucp_worker,
+                  ucp_ep_h *eps, int size, int rank) {
+  auto communicator = std::make_shared<MLCommon::cumlCommunicator>(
+    std::unique_ptr<MLCommon::cumlCommunicator_iface>(
+      new cumlNCCLCommunicator_impl(comm, ucp_worker, eps, size, rank)));
+  handle.getImpl().setCommunicator(communicator);
+}
+
+void inject_comms_py(ML::cumlHandle *handle, ncclComm_t comm, void *ucp_worker,
+                     void *eps, int size, int rank) {
+  ucp_worker_print_info((ucp_worker_h)ucp_worker, stdout);
+
+  ucp_ep_h *new_ep_arr = new ucp_ep_h[size];
+
+  size_t *size_t_ep_arr = (size_t *)eps;
+
+  for (int i = 0; i < size; i++) {
+    size_t ptr = size_t_ep_arr[i];
+    if (ptr != 0) {
+      ucp_ep_h *eps_ptr = (ucp_ep_h *)size_t_ep_arr[i];
+      new_ep_arr[i] = *eps_ptr;
+    } else {
+      new_ep_arr[i] = nullptr;
+    }
+  }
+
+  inject_comms(*handle, comm, (ucp_worker_h)ucp_worker, (ucp_ep_h *)new_ep_arr,
+               size, rank);
+}
+
+void ncclUniqueIdFromChar(ncclUniqueId *id, char *uniqueId) {
+  memcpy(id->internal, uniqueId, NCCL_UNIQUE_ID_BYTES);
+}
+
+/**
+ * @brief Returns a NCCL unique ID as a character array. PyTorch
+ * uses this same approach, so that it can be more easily
+ * converted to a native Python string by Cython and further
+ * serialized to be sent across process & node boundaries.
+ *
+ * @returns the generated NCCL unique ID for establishing a
+ * new clique.
+ */
+void get_unique_id(char *uid) {
+  ncclUniqueId id;
+  ncclGetUniqueId(&id);
+
+  memcpy(uid, id.internal, NCCL_UNIQUE_ID_BYTES);
+}
+
+cumlNCCLCommunicator_impl::cumlNCCLCommunicator_impl(ncclComm_t comm,
+                                                     ucp_worker_h ucp_worker,
+                                                     ucp_ep_h *eps, int size,
+                                                     int rank)
+  : _nccl_comm(comm),
+    _ucp_worker(ucp_worker),
+    _ucp_eps(eps),
+    _size(size),
+    _rank(rank),
+    _next_request_id(0) {}
+
+cumlNCCLCommunicator_impl::~cumlNCCLCommunicator_impl() {}
+
+int cumlNCCLCommunicator_impl::getSize() const { return _size; }
+
+int cumlNCCLCommunicator_impl::getRank() const { return _rank; }
+
+std::unique_ptr<MLCommon::cumlCommunicator_iface>
+cumlNCCLCommunicator_impl::commSplit(int color, int key) const {
+  // Not supported by NCCL
+  ASSERT(
+    false,
+    "ERROR: commSplit called but not supported in this comms implementation.");
+}
+
+void cumlNCCLCommunicator_impl::barrier() const {
+  int *sendbuff, *recvbuff;
+
+  cudaStream_t stream;
+  cudaStreamCreate(&stream);
+
+  CUDA_CHECK(cudaMalloc((void **)&sendbuff, sizeof(int)));
```

Comment:
To avoid the overhead of constant allocation and deallocation in barrier you could make `sendbuff`, `recvbuff` and `stream` members of `cumlNCCLCommunicator_impl`.